### PR TITLE
Convert `product-control` to TypeScript

### LINKED ISSUE #9537: Convert `index.js` to `index.tsx` and replace `propTypes` with TypeScript definitions
## Problem

[`assets/js/editor-components/product-control/index.js`](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/assets/js/editor-components/product-control/index.js) is a JS file and uses `propTypes` definitions.

## Expected outcome

1. Convert the file from `index.js` to `index.tsx`.
2. Replace `propTypes` with TypeScript definitions.

### PR DESCRIPTION
**Pull request by sunyatasattva**
The main aim of this PR was to convert the `product-control` component to TypeScript. However, since the component is coupled with many other parts (it is itself wrapped into 4 HOCs), many refactoring changes were needed to make the types work.

In particular:

1. Many components using the `SearchList` are just passing all sorts of objects to the components with different shapes. Ideally, only the required props for the items would be passed but I understand there are circumstances in which properties of an object might be required. So, I changed to type to allow for a `details` key within the `SearchItem` to receive all sorts of different types and still be strictly typed via generics. Although the generic comes with an untyped, unconstrained default for backwards compatibility, this still required quite a few changes around.
   -  Note that to use the full power of these types, more consumers of the `SearchList` would need to be refactored slightly, though it's not in the scope of this PR. Still, the type safety should be unaffected.
2. Fix `ProductResponseItem` type missing `slug`: it was present in the API response, but not reflected in the type.
3. Improve certain HOC types:
   - The `withSearchedProducts` and `withTransformSingleSelectToMultipleSelect` HOCs are now typed in such a way that it is clear what incoming props they expect and what they inject. 
4. Refactor `ProductAttributeTermControl` for further type safety.

Fixes #9537

#### Other Checks

- [x] This PR has either a `[type]` label or a `[skip-changelog]` label.

#### Automated Tests
* [ ] Changes in this PR are covered by Automated Tests.
  * [ ] Unit tests
  * [ ] E2E tests

#### User Facing Testing

1. Smoke test the `ProductControl` component by using blocks such as the Featured Product or Products by Attribute.

* [ ] Do not include in the Testing Notes 
* [ ] Should be tested by the development team exclusively 

### WooCommerce Visibility

* [x] WooCommerce Core
* [ ] Feature plugin
* [ ] Experimental

### CODE REVIEW

**Comment by sunyatasattva** [comment]
I'd love to ask @thealexandrelara and @kmanijak to check point (3.), in particular the `withTransformSingleSelectToMultipleSelect` HOCs as you had converted and reviewed the TypeScript conversion in #9798. I made some changes to the TypeScript conversion and I want to make sure I didn't miss or misunderstand some intentions you had: it seemed to me that it wasn't correctly typing the expected incoming props (from original component), nor was it providing the type for the wrapper.

But maybe I missed something and I wouldn't want to step on your toes, so please do have a look at that part of this PR 🙏

File: assets/js/editor-components/product-control/index.tsx [hunk 1]
```diff
@@ -3,26 +3,64 @@
...
 import classNames from 'classnames';
-import ExpandableSearchListItem from '@woocommerce/editor-components/expandable-search-list-item/expandable-search-list-item.tsx';
+import ExpandableSearchListItem from '@woocommerce/editor-components/expandable-search-list-item/expandable-search-list-item';
 
 /**
  * Internal dependencies
  */
 import './style.scss';
 
+interface Props {
```
Comment (thealexandrelara) [inline assets/js/editor-components/product-control/index.tsx]: Should this be renamed to `ProductControlProps` to keep consistency with the naming used in other files?

**Review by thealexandrelara — APPROVED** [review]
Great work! It's fantastic to see the usage of generics in those files.

Initially, when I converted that file to TypeScript, it was necessary to add Record<string, unknown> which basically means that any key/value pair is accepted. But with the addition of generics, we can provide more accurate types for blocks using those HOCs.

I left just a small comment, but I'm pre-approving this PR since everything is working as expected and the code looks good.

File: assets/js/editor-components/product-control/index.tsx (hunk 1 above)
Comment (sunyatasattva) [reply assets/js/editor-components/product-control/index.tsx]: Good catch! Definitely! I'll update it!

**Pull request merged**